### PR TITLE
Deprecate Epetra

### LINKED ISSUE #69: Deprecate Epetra
## Description

With current efforts in Trilinos as well as in Baci to deprecate the use of the `Epetra` linear algebra stack, also MIRCO should drop its dependency on `Epetra`. 

As of today, we only rely on some serial dense objects in `Epetra`, which need to be replaced by their counterpart in `Teuchos`. Here's the list of necessary replacements:
- `Epetra_SerialDenseMatrix` --> `Teuchos::SerialDenseMatrix`
- `Epetra_SerialSymDenseMatrix` --> `Teuchos::SerialSymDenseMatrix`
- `Epetra_SerialDenseVector` --> `Teuchos::SerialSymDenseVector`
- `Epetra_SerialDenseSolver` --> `Teuchos::SerialDenseSolver`
- `Epetra_SerialSpdDenseSolver` --> `Teuchos::SerialSpdDenseSolver`

This might bring along to use `Teuchos::RCP` in oder to pass the correct data types into the `Teuchos::SerialDenseSolver` and its variants. This should not alter any test results, just replace an outdated software dependency.

In terms of the build system, we could refine the `find_package()` command for Trilinos, such that it only exposes `Teuchos` to MIRCO, so that we do not accidentally link against Trilinos packages that we do not want to support (anymore). Therefore, one could update the call `find_package()` for Trilinos as follows:

```cmake
find_package(Trilinos REQUIRED COMPONENTS Teuchos)
```

## Possible Solution and Definition of Done

- [x] Replace `Epetra_SerialDense*` with `Teuchos::SerialDense*` where necessary.
- [x] Do not include any `Epetra` headers anymore
- [x] Update `find_package()` of Trilinos to only expose `Teuchos`

## Related Issues and Merge Requests

* Blocks
* Is blocked by
* Follows
* Precedes
* Related to
* Part of
* Composed of

## Additional Information

While Baci still takes some time to complete the `Epetra` deprecation, MIRCO can go forward here. We can even adapt the MIRCO inclusion into Baci to the `Teuchos` interface, so that MIRCO is on the safe side and will not have to be touched, when Baci deprecated `Epetra`.

## Interested Parties

@RShaw026 @NoraHagmeyer

### PR DESCRIPTION
**Pull request by mayrmt**
## Description and Context

- Replace `Epetra` serial dense numerics with `Teuchos` serial dense numerics.
- Expose only the `Teuchos` package to be used within MIRCO.

## Related Issues and Pull Requests

* Closes #69 

## How Has This Been Tested?

Test suite not passing yet. Still WIP.

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers and are longer than one line where appropriate.
- [x] I have added/updated documentation where necessary.

## Interested Parties / Possible Reviewers

@RShaw026 @NoraHagmeyer

### CODE REVIEW

**Comment by mayrmt** [comment]
## Next steps

- Fix failing tests
- Re-evaluate the use of specialized matrices and solvers to exploit symmetry and positive definiteness
- (Maybe) use `Teuchos::SerialDenseVector` for vectors (instead of misusing `Teuchos::SerialDenseMatrix` with just one column)

**Comment by NoraHagmeyer** [comment]
> ## Next steps
> * [ ]  Fix failing tests
> * [ ]  Re-evaluate the use of specialized matrices and solvers to exploit symmetry and positive definiteness
> * [ ]  (Maybe) use `Teuchos::SerialDenseVector` for vectors (instead of misusing `Teuchos::SerialDenseMatrix` with just one column)

Keeping up-to-date with developments in trilinos is a very important step since MIRCO heavily relies on it. Though, if the linear solver is changed to a shared-memory variant (#1) in the near future some of these steps will probably be overkill. As I was not privy to the last "MIRCO - next steps" discussion I can't speak to the plan going forward though. Maybe @isteinbrecher has some thoughts on that as well. 
In any case, this is a valuable contribution, and since the work is already done, my comment is probably not as helpful anymore. In light of #1, reevaluating the speed of the linear solver might not be a priority though.

**Comment by NoraHagmeyer** [comment]
Regarding the tests.. I can't say what fails.. but until now MIRCO has relied on the fact that the matrix is symmetric and the solver actually only ready one half of the matrix.. this might give you problems now

**Comment by isteinbrecher** [comment]
Looks good to me, definitely a good thing to get rid of Epetra. Regarding the `Teuchos::SerialDenseVector`, since this one inherits from `Teuchos::SerialDenseMatrix` the change should be straight forward.

**Comment by mayrmt** [comment]
Tests are passing after exploiting symmetry and positive definiteness as in the Epetra implementation. I also spent a couple of minutes to introduce some `Teuchos::SerialDenseVector`s for vectors.

**Comment by NoraHagmeyer** [comment]
Apparently, the workflow has been disabled for the last 3 months because of inactivity.. all tests should be regularly run and at the very least be run before merging. @RShaw026 Please make sure that the state of MIRCO is well-tested and this does not happen. I hope simply reinstating the workflow does the trick and MR are automatically tested again in the future.

**Comment by mayrmt** [comment]
My comment about failing tests was not related to automated testing, just to failing tests after I made changes to the code locally on my machine.

**Comment by mayrmt** [comment]
I'm done so far with this branch. Please have a look! We can then decide if we merge it as is or if further changes make sense.

**Comment by NoraHagmeyer** [comment]
> My comment about failing tests was not related to automated testing, just to failing tests after I made changes to the code locally on my machine.
> 
> EDIT: In fact, I just was notified about an automated workflow run. So, automated testing seems to work.

I know, it is now running again since I reenabled automatic testing. The tests are meant to be automatically run every time a MR is opened and during a daily scheduled pipeline. MRs are not meant to be approved and merged without a proper run of the workflow. Looking at your MR I was wondering why I couldn't find any workflow.

File: src/mirco_linearsolver.cpp [hunk 1]
```diff
@@ -1,25 +1,29 @@
...
-  int err = solver.SetMatrix(matrix);
+  Teuchos::SerialSpdDenseSolver<int,double> solver;
+  int err = solver.setMatrix(Teuchos::rcpFromRef(matrix));
   if (err != 0)
   {
     std::cout << "Error setting matrix for linear solver (1)";
   }
 
-  err = solver.SetVectors(vector_x, vector_b);
+  err = solver.setVectors(Teuchos::rcpFromRef(vector_x), Teuchos::rcpFromRef(vector_b));
```
Comment (NoraHagmeyer) [inline src/mirco_linearsolver.cpp:19]: Would it make sense to simply create the vectors as RCPs and pass them as such in the first place?

**Review by NoraHagmeyer — APPROVED** [review]
@mayrmt Thanks for the contribution and updating MIRCO in order to use new trilinos developments in the future! Changes look good to me, I just mentioned one small thing that might improve the performance. Otherwise, you are good to merge from my side!

File: src/mirco_linearsolver.cpp (hunk 1 above)
Comment (mayrmt) [reply src/mirco_linearsolver.cpp:19]: That might be true. I'll discuss with @RShaw026 how to do that.

Comment (mayrmt) [reply src/mirco_linearsolver.cpp:19]: I had a quick look. This requires further changes in some other files. We'll probably postpone that to a different PR.

Comment (NoraHagmeyer) [reply src/mirco_linearsolver.cpp:19]: Ok with me.. can you please open an issue to make sure we don't forget?

**Comment by RShaw026** [comment]
@maxfirmbach This PR solves the compile errors I was getting in BACI for using `Epetra_SerialSymDenseMatrix` and `Epetra_SerialSpdDenseSolver` in MIRCO. Therefore, merging this PR.

File: src/mirco_linearsolver.cpp (hunk 1 above)
Comment (RShaw026) [reply src/mirco_linearsolver.cpp:19]: I will open an issue.

**Pull request merged**